src/librustdoc/html/format.rs

```diff
@@ -367,14 +367,20 @@ crate fn print_where_clause<'a, 'tcx: 'a>(
 
         if !f.alternate() {
             clause.push_str("</span>");
-            let padding = "&nbsp;".repeat(indent + 4);
-            clause = clause.replace("<br>", &format!("<br>{}", padding));
-            clause.insert_str(0, &"&nbsp;".repeat(indent.saturating_sub(1)));
-            if !end_newline {
-                clause.insert_str(0, "<br>");
+            let mut br_with_padding = "<br>".to_string();
+            for _ in 0..indent + 4 {
+                br_with_padding.push_str("&nbsp;");
+            }
+            clause = clause.replace("<br>", &br_with_padding);
+
+            let mut left_padding = br_with_padding;
+            left_padding.truncate(if end_newline { 0 } else { "<br>".len() });
+            for _ in 0..indent.saturating_sub(1) {
+                left_padding.push_str("&nbsp;");
             }
+            clause = left_padding + &clause;
```

Comment:
This doesn't seem any easier-to-follow than the previous version. Here's the things that I really don't like about either version:

1. It uses `indent.saturating_sub(1)`. Does this mean `left_padding` has the same value when `indent=0` as it has when `indent=1`? If so, isn't that a bug? If not, why not?
2. This is doing string replacement on a value that was already produced with string concatenation above. Isn't this slow? Why isn't it building the string correctly the first time? I would imagine setting up `br_with_padding` at the beginning of the function and then using it below, instead of doing this `clause.replace()` stuff, would be faster (though, perhaps, not faster-enough to show up in profiling).
3. `clause = left_padding + &clause` and `clause.insert_str(0, left_padding)` are the same thing. One isn't cleaner than the other.

Reply:
1. I'm not familiar with rustdoc's internals, so I'm not sure about this.
2. In this function, the only concatenation to `clause` between the call to `comma_sep` and the string replacement adds six bytes at most (`"&nbsp;"`), so the extra overhead seems insignificant. All direct calls to `comma_sep` are in this file, but tracking down the inputs that are eventually fed into `comma_sep` seems troublesome.
3. Fair. I think it's marginally more self-evident, but I definitely see where you're coming from.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=afd5bfce027f84ae4b54cf29f8995fd6
```
f_alternate: false
end_newline: false
clause: <br>&nbsp;&nbsp; <span class="where">where[COMMA_SEP]</span>

f_alternate: false
end_newline: true
clause: &nbsp;&nbsp; <span class="where fmt-newline">where[COMMA_SEP],&nbsp;</span>

f_alternate: true
end_newline: false
clause:  where[COMMA_SEP]

f_alternate: true
end_newline: true
clause:  where[COMMA_SEP],
```

This doesn't help with the string replacement concern (not sure that can be resolved easily), but this might be useful for simplifying the logic in the bottom half of the function. Even with just two boolean flags, it's quite gnarly.

Reply:
Yeah, a match with four branches would be a great way to clean that thing up. The code might be a little longer, but it would be much, much easier to understand.

Reply:
I previously implemented it with a match expression. Currently, it's two layers of if-else nesting (one if-else expression in each branch) since I've found that it's slightly more amenable to duplicating code across branches. I think it's still readable.